lisp/mastodon-notifications.el

```diff
@@ -0,0 +1,168 @@
+  "Add byline for TOOT with MESSAGE."
+      (concat
+       " "
+       (propertize message 'face 'highlight)
+       " "
+       "You!"))
+
+(defun mastodon-notifications--byline (toot message)
+  "Generate byline from TOOT based on MESSAGE for notification."
+  (let ((id (cdr (assoc 'id toot)))
+        (faved (mastodon-tl--field 'favourited toot))
+        (boosted (mastodon-tl--field 'reblogged toot)))
+    (propertize
+     (concat (propertize "\n | " 'face 'default)
+             (when boosted
+               (format "(%s) " (propertize "B" 'face 'success)))
+             (when faved
+               (format "(%s) " (propertize "F" 'face 'success)))
+             (mastodon-tl--byline-author toot)
+             (mastodon-notifications--byline-concat toot message)
+             (propertize "\n  ------------" 'face 'default))
+     'toot-id id
+     'toot-json toot)))
+
+(defun mastodon-notifications--mention (note)
+  "Format for a `mention' NOTE."
+  (let ((toot (mastodon-tl--field 'status note)))
+    (insert
+     (concat (mastodon-tl--content toot)
+             (mastodon-notifications--byline toot "Mentioned")
+             "\n\n"))))
+
+(defun mastodon-notifications--follow (note)
+  "Format for a `follow' NOTE."
+  (let ((toot (mastodon-tl--field 'status note)))
+  (insert
+   (concat
+    "Congratulations, you have a new follower!"
+    (mastodon-notifications--byline note "Follows")
+    "\n\n"))))
+
+
+(defun mastodon-notifications--favorite(note)
+  "Format for a `favorite' NOTE."
+  (let ((toot (mastodon-tl--field 'status note)))
+    (insert
+     (concat (mastodon-tl--content toot)
+             (mastodon-notifications--byline note "Favorited")
+             "\n\n"))))
+
+(defun mastodon-notifications--reblog (note)
+  "Format for a `boost' NOTE."
+  (let ((toot (mastodon-tl--field 'status note)))
+    (insert
+     (concat (mastodon-tl--content toot)
+             (mastodon-notifications--byline note "Boosted")
+             "\n\n"))))
+
+(defun mastodon-notifications--caller (type note)
+  "Call the apprpriate function bassed on notification TYPE.
+
+It then processes NOTE."
+  (cond ((equal type "mention") (mastodon-notifications--mention note))
+        ((equal type "follow") (mastodon-notifications--follow note))
+        ((equal type "favourite") (mastodon-notifications--favorite note))
+        ((equal type "reblog") (mastodon-notifications--reblog note))))
+
+(defun mastodon-notifications--note (note)
+  "Filters NOTE for those listed in `mastodon-notifications--types'.
+
+  It then passes the results to the `caller'"
+  (let ((type (mastodon-tl--field 'type note)))
+    (when (member type mastodon-notifications--types)
+      (mastodon-notifications--caller type note))))
+
+(defun mastodon-notifications--notifications (json)
+  "Format JSON in Emacs buffer."
+  (mapcar #'mastodon-notifications--note json)
+  (replace-regexp "\n\n\n | " "\n | " nil (point-min) (point-max)))
```

Comment:
I've never got to the bottom of this - why do we first insert these newlines only to later remove them again?